test/Microsoft.Framework.DependencyInjection.Tests/ServiceCollectionExtensionTest.cs

```diff
@@ -276,87 +285,154 @@ public void TryAddWithEnumerableReturnsFalseIfNoneAdded()
-            Assert.Single(collection, sd => sd.ImplementationType == typeof(FakeService));
+            Assert.True(collection.TryAddEnumerable(ds));
+            var descriptor = Assert.Single(collection);
+            Assert.Equal(typeof(IFakeService), descriptor.ServiceType);
+            Assert.Null(descriptor.ImplementationInstance);
+            Assert.Equal(ServiceLifetime.Transient, descriptor.Lifetime);
         }
 
         [Fact]
-        public void TryAddEnumerable_AllowsMultipleRegistrations_WithInstance()
+        public void TryAddEnumerable_WithEnumerable_ReturnsFalseIfNoneAdded()
         {
             // Arrange
             var collection = new ServiceCollection();
-            collection.Add(ServiceDescriptor.Instance<IFakeService>(new FakeService()));
+            collection.AddSingleton<IFakeService, FakeService>();
 
             // Act
-            var result = collection.TryAddEnumerable(ServiceDescriptor.Transient<IFakeService, FakeService>());
+            var ds = new ServiceDescriptor[] {
+                ServiceDescriptor.Transient<IFakeService, FakeService>(),
+                ServiceDescriptor.Transient<IFakeService, FakeService>()
+            };
 
             // Assert
-            Assert.True(result);
+            Assert.False(collection.TryAddEnumerable(ds));
+            var descriptor = Assert.Single(collection);
+            Assert.Equal(typeof(IFakeService), descriptor.ServiceType);
+            Assert.Null(descriptor.ImplementationInstance);
+            Assert.Equal(ServiceLifetime.Singleton, descriptor.Lifetime);
+        }
 
-            Assert.Single(collection, sd => sd.ImplementationInstance != null);
-            Assert.Single(collection, sd => sd.ImplementationType == typeof(FakeService));
+        public static TheoryData TryAddEnumerableInvalidImplementationTypeData
+        {
+            get
+            {
+                var serviceType = typeof(IFakeService);
+                var implementationType = typeof(FakeService);
+                var objectType = typeof(object);
+
+                return new TheoryData<ServiceDescriptor, Type, Type>
+                {
+                    { ServiceDescriptor.Transient<IFakeService>(s => new FakeService()), serviceType, serviceType },
+                    { ServiceDescriptor.Transient(serviceType, s => new FakeService()), serviceType, objectType },
+
+                    { ServiceDescriptor.Scoped<IFakeService>(s => new FakeService()), serviceType, serviceType },
+                    { ServiceDescriptor.Scoped(serviceType, s => new FakeService()), serviceType, objectType },
+
+                    { ServiceDescriptor.Singleton<IFakeService>(s => new FakeService()), serviceType, serviceType },
+                    { ServiceDescriptor.Singleton(serviceType, s => new FakeService()), serviceType, objectType },
+                };
+            }
         }
 
-        [Fact]
-        public void TryAddEnumerable_ThrowsForNonType()
```

Comment:
Maybe we add a test similar to this one that was removed for when `ImplementationType`, `ImplementationInstance` and `ImplementionFactory` are all null.